### PR TITLE
Deduplicate violations in the source space

### LINKED ISSUE #4040: Deduplicate violations in the same position
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

When linting jinja files with loops we get multiple output violations for each time around the loop. e.g.

```sql
select
    a,
    {% for val in [1, 2, 3, 4, 5, 6] %}
        d+ {{ val }},
    {% endfor %}
    b
```

we get

```
== [test.sql] FAIL
L:   4 | P:  10 | L006 | Missing whitespace before +
L:   4 | P:  10 | L006 | Missing whitespace before +
L:   4 | P:  10 | L006 | Missing whitespace before +
L:   4 | P:  10 | L006 | Missing whitespace before +
L:   4 | P:  10 | L006 | Missing whitespace before +
L:   4 | P:  10 | L006 | Missing whitespace before +
L:   7 | P:   1 | L001 | Unnecessary trailing whitespace.
```

The duplicated `Missing whitespace` isn't helpful for the user. Regardless of whether we keep them in the background (perhaps we should), they shouldn't be shown to the user here because we're showing the same issue multiple times.

### Use case

CLI linting

### Dialect

all

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This resolves #4040.

Duplicate violations (e.g. from multiple passes through a jinja loop) confuse the user and also end up just being deduplicated later. This filters them out early.

### CODE REVIEW

File: test/core/linter_test.py [hunk 1]
```diff
@@ -406,8 +406,6 @@ def test__linter__empty_file():
         (
             False,
             [
-                ("L006", 3, 16),
-                ("L006", 3, 16),
                 ("L006", 3, 16),
```
Comment (barrywhart) [inline test/core/linter_test.py:409]: Even after deduplication, it looks like there are still 2 occurrences of:
```
("L006", 3, 16),
```

Any idea why?

Comment (alanmcruickshank) [reply test/core/linter_test.py:412]: Ah - yeah I can see this is confusing. The rules we're checking is L006, which checks for whitespace *before* and *after* the operator. I've added a comment to explain this.

Comment (alanmcruickshank) [reply test/core/linter_test.py:412]: Deduplication means that we only get 2 rather than 4.

File: src/sqlfluff/core/linter/linted_file.py [hunk 2]
```diff
@@ -54,18 +54,43 @@ def check_tuples(self, raise_on_non_linting_violations=True) -> List[CheckTuple]
...
+        new_violations = []
+        dedupe_buffer = set()
+        for v in violations:
+            signature = v.source_signature()
+            if signature not in dedupe_buffer:
+                new_violations.append(v)
+                dedupe_buffer.add(signature)
+            else:
+                linter_logger.debug("Removing duplicate source violation: %s", v)
+        return new_violations
```
Comment (pwildenhain) [inline src/sqlfluff/core/linter/linted_file.py:92]: 👨🏻‍🍳 👌🏻

**Review by pwildenhain — APPROVED** [review]
✅

**Pull request merged**